sdk/search/Azure.Search/src/Models/IndexDocumentsAction{T}.cs

```diff
@@ -0,0 +1,216 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.Text.Json;
+using Azure.Core;
+
+#pragma warning disable SA1402 // File may only contain a single type
+
+namespace Azure.Search.Models
+{
+    // Hide the untyped IndexAction
+    [CodeGenSchema("IndexAction")]
+    internal partial class IndexAction { }
+
+    /// <summary>
+    /// Represents an index action that operates on a document.
+    /// </summary>
+    /// <typeparam name="T">
+    /// The .NET type that maps to the index schema. Instances of this type can
+    /// be retrieved as documents from the index.
+    /// </typeparam>
+    public class IndexDocumentsAction<T> : IUtf8JsonSerializable
+    {
+        /// <summary>
+        /// The operation to perform on a document in an indexing batch.
+        /// </summary>
+        public IndexActionType ActionType { get; private set; }
+
+        /// <summary>
+        /// The document to index.
+        /// </summary>
+        public T Document { get; private set; }
+
+        /// <summary>
+        /// Private constructor to prevent direct instantiation.
+        /// </summary>
+        private IndexDocumentsAction() { }
+
+        /// <summary>
+        /// Initializes a new instance of the IndexDocumentsAction class.
+        /// </summary>
+        /// <param name="type">
+        /// The operation to perform on the document.
+        /// </param>
+        /// <param name="doc">The document to index.</param>
+        public IndexDocumentsAction(IndexActionType type, T doc)
+        {
+            Debug.Assert(Enum.IsDefined(typeof(IndexActionType), type));
+
+            ActionType = type;
+            Document = doc;
+        }
+
+        /// <summary>
+        /// Serialize the document write action.
+        /// </summary>
+        /// <param name="writer">The JSON writer.</param>
+        void IUtf8JsonSerializable.Write(Utf8JsonWriter writer)
+        {
+            Debug.Assert(writer != null);
+
+            writer.WriteStartObject();
+
+            writer.WritePropertyName("@search.action");
+            writer.WriteStringValue(ActionType.ToSerialString());
+
+            // TODO: XXXXX - Investigate a more efficient way of injecting properties into user models with System.Text.Json
+
+            // HACK: Serialize the user's model, parse it, and then write each
+            // of its properties as if they were our own.
+            string json = JsonSerializer.Serialize<T>(Document, JsonExtensions.SerializerOptions);
+            using JsonDocument nested = JsonDocument.Parse(json);
+            if (nested.RootElement.ValueKind == JsonValueKind.Object)
```

Comment:
The REST API will reject any document that isn't an object, so `IndexDocumentsAction<int>` is not meaningful. That said, constraining `T` to classes removes the ability of customers to use structs as documents. We supported that in Track 1, but I don't have any data to show how many customers might be using that capability.

Reply:
(I've ported some of those unit tests and verified that structs are working happily in Track 2 as well.)

Reply:
> The REST API will reject any document that isn't an object

When you say object, do you mean class/struct in .NET, or anything that can written as a {...} in JSON?

What about a JSON array, List<int> or any other IEnumerable that end up written out as [...]?

Reply:
I mean JSON objects. Here's what the payload looks like: https://docs.microsoft.com/en-us/rest/api/searchservice/addupdate-or-delete-documents#examples

Reply:
Great. Thanks for clarifying. Presumably, given this is a public type, someone could use `IndexDocumentsAction<T>` outside the context of the search service, and then hit this edge case (by just newing it up with arbitrary serializable Ts), but that is likely an unsupported case. Is it still worth guarding against that and throwing, rather than silently ignoring? 

And if not, should we just remove that `ValueKind == Object` check all together and in the off-chance someone uses it incorrectly, let EnumerateObject throw instead?

To me, the fact that we are checking threw me off. If it is expected to be a JSON object, just don't check.

Reply:
Removing the ValueKind check so it'll throw instead.